In [1]:
import pandas as pd
import sys
from pathlib import Path

In [2]:
# forcing Jupyter to show output for all operations in a cell and not just last one


from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"

#  InteractiveShell.ast_node_interactivity = "last_expr" will revert to original

In [3]:
%%HTML
<style type="text/css">
    table.dataframe td, table.dataframe th {
        border-style: solid;
    }
</style

In [4]:
#pd.set_option('precision', 2) to display 2 decimal digits in output( default setting is 6 digits)

# setting column width
from IPython.core.display import HTML
HTML("<style>.rendered_html th {max-width: 120px;}</style>")

In [5]:
! ls

04-Pandas_Instructions_HeroesOfPymoli_Resources_purchase_data.csv
Pymoli_notebook.ipynb


In [6]:
# set path to read input file

data_folder = Path("") # input file in current directory no no folder path is empty
data_file = "04-Pandas_Instructions_HeroesOfPymoli_Resources_purchase_data.csv"

file_to_read = data_folder / data_file

In [7]:
# read from csv file
df = pd.read_csv(file_to_read)
df.head()
df.info()

Purchase ID             SN  Age Gender  Item ID  \
0            0        Lisim78   20   Male      108   
1            1    Lisovynya38   40   Male      143   
2            2     Ithergue48   24   Male       92   
3            3  Chamassasya86   24   Male      100   
4            4      Iskosia90   23   Male      131   

                                   Item Name  Price  
0  Extraction, Quickblade Of Trembling Hands   3.53  
1                          Frenzied Scimitar   1.56  
2                               Final Critic   4.88  
3                                Blindscythe   3.27  
4                                       Fury   1.44

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 780 entries, 0 to 779
Data columns (total 7 columns):
Purchase ID    780 non-null int64
SN             780 non-null object
Age            780 non-null int64
Gender         780 non-null object
Item ID        780 non-null int64
Item Name      780 non-null object
Price          780 non-null float64
dtypes: float64(1), int64(3), object(3)
memory usage: 42.7+ KB


### __Player Count - Total number of players__

In [8]:
# displaying count of total players as dataframe
total_players = df.loc[:, 'SN'].nunique() # returns an integer

# printing as a table by converting to DataFrame
pd.DataFrame({'Total Players': total_players}, index =[0])


Total Players
0            576

### Purchasing Analysis (Total)

In [9]:
# performing basic purchase analysis 
number_unique_items = df.loc[:, 'Item ID'].nunique()
average_purchase_price =  round(df.loc[:, 'Price'].mean(), 2)
total_number_of_purchases = len(df )
total_revenue = df.loc[:, 'Price'].sum()

# convert to a dataframe for tabular print
purchasing_analysis = pd.DataFrame({'Number of Unique Items':number_unique_items, 
                                    'Average Purchase Price': "$"+str(average_purchase_price), 
                                    'Total Number of Purchases':total_number_of_purchases, 
                                    'Total Revenue':"$"+str(total_revenue)}, index= [0])
purchasing_analysis

Number of Unique Items Average Purchase Price  Total Number of Purchases  \
0                     183                  $3.05                        780   

  Total Revenue  
0      $2379.77

### Gender Demographics

__using dictionary and defining indexes__ to prepare dataframe object

In [10]:
total_player_count = df.loc[:'Gender'].nunique()['SN']
count_male_player = df.loc[df.loc[:, 'Gender'] == 'Male', :].nunique()['SN']
count_female_player = df.loc[df.loc[:, 'Gender'] == 'Female', :].nunique()['SN']
count_other_player = df.loc[df.loc[:, 'Gender'] == 'Other / Non-Disclosed', :].nunique()['SN']

percentage_male = round(100*count_male_player/total_player_count, 2)
percentage_female = round(100*count_female_player/total_player_count, 2)
percentage_others = round(100*count_other_player/total_player_count, 2)

# aggregating above data into a dictionary to finally display in dataframe

dict = {'Percentage': [percentage_male, percentage_female,percentage_others ], 'Count':[count_male_player, count_female_player, count_other_player] }

In [11]:
# converting dictionary to a dataframe
gender_demo = pd.DataFrame(dict, index = ['Male', 'Female', 'Other / Non-Disclosed'])
gender_demo

Percentage  Count
Male                        84.03    484
Female                      14.06     81
Other / Non-Disclosed        1.91     11

### Purchasing Analysis by Gender
__using lists and defining both indexes and columns__ to prepare dataframe object

In [12]:
# a few of below commands return pd.series data object which can be accessed like a list. 
#To get a specific value use [<integer index of the list item >] / or[<name index of the list item>]. using names

total_purchase_count_male = df.loc[df['Gender'] == 'Male', :].count()['SN']
average_purchase_price_male =  df.loc[df.loc[:,'Gender'] =='Male', ['Price']].mean()['Price']
total_purchase_value_male = df.loc[df.loc[:,'Gender'] =='Male', ['Price']].sum()['Price']
number_of_unique_male_players = df.loc[df.loc[:,'Gender'] =='Male', :].nunique()['SN']

total_purchase_count_female = df.loc[df['Gender'] == 'Female', :].count()['SN']
average_purchase_price_female =  df.loc[df.loc[:,'Gender'] =='Female', ['Price']].mean()['Price']
total_purchase_value_female = df.loc[df.loc[:,'Gender'] =='Female', ['Price']].sum()['Price']
number_of_unique_female_players = df.loc[df.loc[:,'Gender'] =='Female', :].nunique()['SN']

total_purchase_count_other = df.loc[df['Gender'] == 'Other / Non-Disclosed', :].count()['SN']
average_purchase_price_other =  df.loc[df.loc[:,'Gender'] =='Other / Non-Disclosed', ['Price']].mean()['Price']
total_purchase_value_other = df.loc[df.loc[:,'Gender'] =='Other / Non-Disclosed', ['Price']].sum()['Price']
number_of_unique_other_players = df.loc[df.loc[:,'Gender'] =='Other / Non-Disclosed', :].nunique()['SN']

In [13]:
# creating 3 lists as argument for dataframe objects
# then creating data frame with indexs set as 'Male' and 'Female'
# used conversion to string to display $ sign in relevant columns

list_male_stats = [total_purchase_count_male, "$"+str(round(average_purchase_price_male,2)),
                         "$"+str(total_purchase_value_male), 
                         "$"+str(round(total_purchase_value_male/number_of_unique_male_players,2))]

list_female_stats = [total_purchase_count_female, "$"+str(round(average_purchase_price_female,2)),
                         "$"+str(total_purchase_value_female), 
                         "$"+str(round(total_purchase_value_female/number_of_unique_female_players,2))]

list_other_stats = [total_purchase_count_other, "$"+str(round(average_purchase_price_other,2)),
                         "$"+str(total_purchase_value_other), 
                         "$"+str(round(total_purchase_value_other/number_of_unique_other_players,2))]


purchasing_analysis = pd.DataFrame([list_male_stats,list_female_stats, list_other_stats],  
                                index = ['Male', 'Female', 'Other / Non-Disclosed'], columns = ['Purchase Count', 'Average Purchase price',
                                         'Total Purchase Value', 'Average Purchase Total per Person' ] )
purchasing_analysis

Purchase Count Average Purchase price  \
Male                              652                  $3.02   
Female                            113                   $3.2   
Other / Non-Disclosed              15                  $3.35   

                      Total Purchase Value Average Purchase Total per Person  
Male                              $1967.64                             $4.07  
Female                             $361.94                             $4.47  
Other / Non-Disclosed               $50.19                             $4.56

### Age Demographics
Option1: Without Pivot Table

In [14]:
df['Age'].max()

45

In [15]:
bins = [0,10, 14, 19, 24, 29, 34, 39, 40]
age_labels = ['0-10','10-14', '15-19', '20-24', '25-29', '30-34', '35-39', '40+']
df['Age Bracket'] = pd.cut (df['Age'], bins, labels= age_labels) # assignment adds new column, hence not needing 'inplace'

In [16]:
# pandas accepts  dictionary as 1 argument while creating a dataframe
#creating a dictionary for age demographics (pertaining to columns of dataframe)
# index will be passed on separately as age-labels while creating dataframe

dict_age_demo = {'Purchase Count':[],'Total Purchase Value':[], 'Average Purchase Price': [], 'Average Purchase Total per Person by Age Group': []  } 

for label in age_labels:
    count = (df.loc[df['Age Bracket'] == label, :].count()['Age Bracket'])
    unique_count = (df.loc[df['Age Bracket'] == label, :].nunique()['SN'])
    total = (df.loc[df['Age Bracket'] == label, :].sum()['Price'])
    average_price = total/count
    average_purchase_total_per_person = total/unique_count
    dict_age_demo['Purchase Count'].append (count)
    dict_age_demo['Total Purchase Value'].append(total)
    dict_age_demo['Average Purchase Price'].append(average_price)
    dict_age_demo['Average Purchase Total per Person by Age Group'].append(average_purchase_total_per_person)
# dict1.update((label, (df.loc[df['Age Bracket'] == label, :].count()['Age Bracket']))

In [17]:
df_age_demo= pd.DataFrame(dict_age_demo, index = age_labels)
df_age_demo


Purchase Count  Total Purchase Value  Average Purchase Price  \
0-10               32                108.96                3.405000   
10-14              19                 50.95                2.681579   
15-19             136                412.89                3.035956   
20-24             365               1114.06                3.052219   
25-29             101                293.00                2.900990   
30-34              73                214.00                2.931507   
35-39              41                147.67                3.601707   
40+                 6                 16.71                2.785000   

       Average Purchase Total per Person by Age Group  
0-10                                         4.540000  
10-14                                        3.396667  
15-19                                        3.858785  
20-24                                        4.318062  
25-29                                        3.805195  
30-34                                        4.115385  
35-39                                        4.763548  
40+                                          3.342000

In [18]:
# Formatting output 
df_age_demo['Total Purchase Value'] = df_age_demo['Total Purchase Value'].map('${:,.2f}'.format)
df_age_demo['Average Purchase Price'] = df_age_demo['Average Purchase Price'].map('${:,.2f}'.format)
df_age_demo['Average Purchase Total per Person by Age Group'] = df_age_demo['Average Purchase Total per Person by Age Group'].map('${:,.2f}'.format)
df_age_demo

Purchase Count Total Purchase Value Average Purchase Price  \
0-10               32              $108.96                  $3.40   
10-14              19               $50.95                  $2.68   
15-19             136              $412.89                  $3.04   
20-24             365            $1,114.06                  $3.05   
25-29             101              $293.00                  $2.90   
30-34              73              $214.00                  $2.93   
35-39              41              $147.67                  $3.60   
40+                 6               $16.71                  $2.79   

      Average Purchase Total per Person by Age Group  
0-10                                           $4.54  
10-14                                          $3.40  
15-19                                          $3.86  
20-24                                          $4.32  
25-29                                          $3.81  
30-34                                          $4.12  
35-39                                          $4.76  
40+                                            $3.34

### Age Demographics
Option2: Using Pivot Table (Much Simpler...)

In [19]:
#DataFrame.pivot_table(values=None, index=None, columns=None, aggfunc='mean', fill_value=None, margins=False, dropna=True, margins_name='All')[source]¶
df_pivot = df.pivot_table( values = ['Price', 'SN'], index='Age Bracket',  aggfunc={'Price':['count','mean', 'sum'], 'SN':'nunique'})
df_pivot

Price                         SN
            count      mean      sum nunique
Age Bracket                                 
0-10           32  3.405000   108.96      24
10-14          19  2.681579    50.95      15
15-19         136  3.035956   412.89     107
20-24         365  3.052219  1114.06     258
25-29         101  2.900990   293.00      77
30-34          73  2.931507   214.00      52
35-39          41  3.601707   147.67      31
40+             6  2.785000    16.71       5

In [20]:
# adding a column for Average Purchase Total per Person
df_pivot.columns = ['Purchase Count', 'Average Purchase Price', 'Total Purchase Value', 'Number of Players']

# updating the new column with calculated average price per person
df_pivot['Average Purchase Total per Person'] = df_pivot['Total Purchase Value']/df_pivot['Number of Players']

In [21]:
# formatting the dataframe with $ sign and decimal upto 2 places
df_pivot['Total Purchase Value'] = df_pivot['Total Purchase Value'].map('${:,.2f}'.format)
df_pivot['Average Purchase Price'] = df_pivot['Average Purchase Price'].map('${:,.2f}'.format)
df_pivot['Average Purchase Total per Person'] = df_pivot['Average Purchase Total per Person'].map('${:,.2f}'.format)
df_pivot

Purchase Count Average Purchase Price Total Purchase Value  \
Age Bracket                                                               
0-10                     32                  $3.40              $108.96   
10-14                    19                  $2.68               $50.95   
15-19                   136                  $3.04              $412.89   
20-24                   365                  $3.05            $1,114.06   
25-29                   101                  $2.90              $293.00   
30-34                    73                  $2.93              $214.00   
35-39                    41                  $3.60              $147.67   
40+                       6                  $2.79               $16.71   

             Number of Players Average Purchase Total per Person  
Age Bracket                                                       
0-10                        24                             $4.54  
10-14                       15                             $3.40  
15-19                      107                             $3.86  
20-24                      258                             $4.32  
25-29                       77                             $3.81  
30-34                       52                             $4.12  
35-39                       31                             $4.76  
40+                          5                             $3.34

### Top Spenders

In [22]:
# Groupby player name and sum all columns
df_spenders = df.groupby(by='SN').sum()
df_spenders.head()
# return object of groupby will have 'SN' as indexes



Purchase ID  Age  Item ID  Price
SN                                            
Adairialis76          467   16      123   2.28
Adastirin33           142   35      175   4.48
Aeda94                388   17      128   4.91
Aela59                 28   21      119   4.32
Aelaria33             630   23      171   1.79

In [23]:
# creating 3 dataframe objects from groupby function 

df_spenders = df.groupby(by='SN').sum()
df_popular = df.groupby(by='Item ID').count()
df_profitable = df.groupby(by='Item ID').sum()


In [24]:
# Option 1: sorting dataframe objects to get top-5 rows on top
df_spenders = df_spenders.sort_values(by= 'Price', ascending = False)
df_popular = df_popular.sort_values(by= 'Item Name', ascending = False)
df_profitable = df_profitable.sort_values(by= 'Price', ascending = False)

# print top-5 rows to verify
df_spenders.head(5)
df_popular.head(5)
df_profitable.head(5)


Purchase ID  Age  Item ID  Price
SN                                           
Lisosia93           1630  125      451  18.96
Idastidru52         1999   96      527  15.45
Chamjask73          1306   66      339  13.83
Iral74              2285   84      518  13.62
Iskadarya95          713   60      321  13.10

Purchase ID  SN  Age  Gender  Item Name  Price  Age Bracket
Item ID                                                             
178               12  12   12      12         12     12           12
145                9   9    9       9          9      9            9
108                9   9    9       9          9      9            9
82                 9   9    9       9          9      9            9
19                 8   8    8       8          8      8            8

Purchase ID  Age  Price
Item ID                         
178             4960  287  50.76
82              2761  215  44.10
145             4345  187  41.22
92              3773  195  39.04
103             3947  178  34.80

Now we will prepare the results in dictionaries and convert to dataframes in below cells

In [25]:
#creating lists of indexes for top-5 items
top5_spenders_list = [df_spenders.index[i] for i in range (5)]
top5_popular_list = [df_popular.index[i] for i in range (5)]
top5_profitable_list = [df_profitable.index[i] for i in range (5)]

# or rather than sorting the groupby result, we can simply use pandas native method ".largest" as shown below
#top_spenders_list = df_spenders['Price'].nlargest(5).index.tolist()

In [26]:
# make dataframe for top5 spenders
dict2 = {'SN':[], 'Purchase Count': [], 'Average Purchase': [], 'Total Purchase Value':[]}
for i in range(5):
    name = df_spenders.index[i]
    count = df.loc[df['SN'] == name, :].count()['Price']
    total_price = df.loc[df['SN'] == name, :].sum()['Price']
    average_price = total_price/count

    
    dict2['SN'].append(name)
    dict2['Purchase Count'].append(count)
    dict2['Average Purchase'].append(average_price)
    dict2['Total Purchase Value'].append(total_price)
    
top5_spenders = pd.DataFrame(dict2)
top5_spenders.set_index('SN', inplace = True)
top5_spenders['Average Purchase'] = top5_spenders['Average Purchase'].map('${:,.2f}'.format)
top5_spenders['Total Purchase Value'] = top5_spenders['Total Purchase Value'].map('${:,.2f}'.format)
top5_spenders


Purchase Count Average Purchase Total Purchase Value
SN                                                               
Lisosia93                 5            $3.79               $18.96
Idastidru52               4            $3.86               $15.45
Chamjask73                3            $4.61               $13.83
Iral74                    4            $3.40               $13.62
Iskadarya95               3            $4.37               $13.10

In [27]:
# make dataframe for top5 popular items
# creating a dictionary structure with empty lists as present values
dict3 = {'Item ID':[],'Item Name':[], 'Purchase Count': [], 'Item Price': [], 'Total Purchase Value':[]}

def item_name(ID): # function to return Item name against an item ID . df.at or df.loc both can be used
    for i in range (len(df)):
        if df.at[i, 'Item ID'] == ID:
            return df.at[i, 'Item Name']
            break
        else:
            pass

def item_price(ID): # function to return Item unit price against an item ID 
    for i in range (len(df)):
        if df.at[i, 'Item ID'] == ID:
            return df.at[i, 'Price']
            break
        else:
            pass

for i in range(5): # Getting particulars for top-5 popular items and store in a dictionary
    ID = df_popular.index[i]
    
    # using original dataframe to get required details for all popular items 
    name = item_name(ID)
    count = df.loc[df['Item ID'] == ID, :].count()['Price']
    itemized_price = item_price(ID)
    total_price = df.loc[df['Item ID'] == ID, :].sum()['Price']
        
    dict3['Item ID'].append(ID)
    dict3['Item Name'].append(name)
    dict3['Purchase Count'].append(count)
    dict3['Item Price'].append(itemized_price)
    dict3['Total Purchase Value'].append(total_price)
    
    
top5_popular = pd.DataFrame(dict3)
top5_popular.set_index(['Item ID' , 'Item Name'], inplace = True)
top5_popular['Item Price'] = top5_popular['Item Price'].map('${:,.2f}'.format)
top5_popular['Total Purchase Value'] = top5_popular['Total Purchase Value'].map('${:,.2f}'.format)

top5_popular

Purchase Count  \
Item ID Item Name                                                      
178     Oathbreaker, Last Hope of the Breaking Storm              12   
145     Fiery Glass Crusader                                       9   
108     Extraction, Quickblade Of Trembling Hands                  9   
82      Nirvana                                                    9   
19      Pursuit, Cudgel of Necromancy                              8   

                                                     Item Price  \
Item ID Item Name                                                 
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
145     Fiery Glass Crusader                              $4.58   
108     Extraction, Quickblade Of Trembling Hands         $3.53   
82      Nirvana                                           $4.90   
19      Pursuit, Cudgel of Necromancy                     $1.02   

                                                     Total Purchase Value  
Item ID Item Name                                                          
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
145     Fiery Glass Crusader                                       $41.22  
108     Extraction, Quickblade Of Trembling Hands                  $31.77  
82      Nirvana                                                    $44.10  
19      Pursuit, Cudgel of Necromancy                               $8.16

In [28]:
# make dataframe for top5 profitable items
# creating a dictionary structure 
dict4 = {'Item ID':[],'Item Name':[], 'Purchase Count': [], 'Item Price': [], 'Total Purchase Value':[]}

for i in range(5): # Getting particulars for top-5 popular items and storint in dictionary
    ID = df_profitable.index[i]
    name = item_name(ID)# function already defined above
    count = df.loc[df['Item ID'] == ID, :].count()['Price']
    itemized_price = item_price(ID) # functional already defined above
    total_price = df.loc[df['Item ID'] == ID, :].sum()['Price']
        
    dict4['Item ID'].append(ID)
    dict4['Item Name'].append(name)
    dict4['Purchase Count'].append(count)
    dict4['Item Price'].append(itemized_price)
    dict4['Total Purchase Value'].append(total_price)
    
    
top5_profitable = pd.DataFrame(dict4)
top5_profitable.set_index(['Item ID' , 'Item Name'], inplace = True)
top5_profitable['Item Price'] = top5_profitable['Item Price'].map('${:,.2f}'.format)
top5_profitable['Total Purchase Value'] = top5_profitable['Total Purchase Value'].map('${:,.2f}'.format)

top5_profitable

Purchase Count  \
Item ID Item Name                                                      
178     Oathbreaker, Last Hope of the Breaking Storm              12   
82      Nirvana                                                    9   
145     Fiery Glass Crusader                                       9   
92      Final Critic                                               8   
103     Singed Scalpel                                             8   

                                                     Item Price  \
Item ID Item Name                                                 
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
82      Nirvana                                           $4.90   
145     Fiery Glass Crusader                              $4.58   
92      Final Critic                                      $4.88   
103     Singed Scalpel                                    $4.35   

                                                     Total Purchase Value  
Item ID Item Name                                                          
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
82      Nirvana                                                    $44.10  
145     Fiery Glass Crusader                                       $41.22  
92      Final Critic                                               $39.04  
103     Singed Scalpel                                             $34.80

# Summary: Three observable trends based on the data

### 1. Majority of the players are male (More than 84% ). 
### 2. Age group vs purchase count has a normal distribution centered around players in age group 20-24. 
### 3. in terms of per person spent, females players are spending approx. 10% higher than male players. Also per peson spent is highest for age group 35-39, follwed by age group 0-10. 